### PR TITLE
Gursaanj/resolver tool

### PR DESCRIPTION
**Pull request by Gursaanj**
**Description**
At the current moment each field with the `[Resolve] `attribute will have an altered property drawer with 4 buttons (Resolve via children, via parent, via self, and the clear current field).

It currently works for normal object reference fields, fields as a part of arrays/lists as well as fields within structs (or as arrays/Lists defined in structs) but I am pretty sure it will break with more complex nesting of fields (i.e arrays of structs that have arrays) but I will worry about that once we finalize if we like this tool/find it useful.

Also for a strange bug related to property drawers (I can explain in detail next time we speak to each other) Don't put the `[Resolve]`  Attribute on a field for a struct. Rather, put in on the fields declared within the struct (as can be seen in ResolveTestScript.cs ).
When you have the time play around in the `ResolveTestScene `scene

**Screenshots / GIFs**
image

image

### CODE REVIEW

**Comment by m4rtint** [comment]
The [Resolve] Currently doesn't work outside of `_utility` because it's outside the asmdef file. 
you need to either put the PropertyAttribute folder into the Scripts folder, or create a separate asmdef. 

I also would like to see how Odin does their asmdef

**Comment by m4rtint** [comment]
In terms of inactive components, this is how Odin sets up their attributes. 
idk if it's nice to follow something similiar for inactive components. 
Imo, it would be nice to default to Including Inactive

Screen Shot 2021-12-31 at 12 47 50 PM

**Comment by m4rtint** [comment]
Disable it if it's a scriptableObject?

we can leave it to iteration 2

Screen Shot 2021-12-31 at 12 57 55 PM

**Comment by m4rtint** [comment]
I'm trying to use Resolver in structs, just like how you did it, but it's not working. 

I read that. you don't want to allow Resolver on structs? Is that the long term goal? 

Screen Shot 2021-12-31 at 1 03 16 PM
Screen Shot 2021-12-31 at 1 03 40 PM

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:32]: Are all these needed or...

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:96]: Can you update this comment. 
Comments are dangerous cuz if doesn't change. 👻

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:69]: How do you include Inactive currently? are we able to change it?

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:136]: In terms of cleaning up, you can always abstract this out into separate methods so it's more readable

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:155]: NiT: maybe rename this, since it's confusing. like `mbParent`

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:238]: >"|Clear Component"
Do you need the "|"

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:235]: "Get Component(s) from Children"

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:281]: Can you make a ticket for this as a reminder. and link the ticket here

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (unoctium1) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:69]: If I'm reading it right there's an optional constructor and it defaults to true, so if you don't want to include inactive children you'd go `[Resolve(false)]`

**Comment by unoctium1** [comment]
@m4rtint my understanding is that your mistake with the struct is putting the [Resolve] attribute on the struct itself, you'd need it on the fields within the struct only. Imo this makes sense, since structs can't ever be complete scene references themselves

**Comment by m4rtint** [comment]
> @m4rtint my understanding is that your mistake with the struct is putting the [Resolve] attribute on the struct itself, you'd need it on the fields within the struct only. Imo this makes sense, since structs can't ever be complete scene references themselves

But in the image I posted, I only set the resolve on an int..
 Which I realize doesn't make sense. But shouldn't really show the error.

**Comment by unoctium1** [comment]
@m4rtint its the `[Resolve][SerializeField] private SomethingElse _something` line that's giving the error. The `int` field and the struct itself is set up correctly I believe

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:69]: Yup exactly, it will default to search inactive gameobjects. and you can toggle that off with the optional parameter

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:136]: Yeah in reality, both `Components` and `GameObjects` derive from `Object` so I shouldn't really need two separate implementations. And instead of using `GetInstanceID()` to get the ID for ObjectReference I could use `GetHashCode` which Object class implements.

Could possibly fix the `ScriptableObject` derivation issue as well now that I think about it

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:238]: Yeah. To be honest I have no idea why haha!

https://docs.unity3d.com/ScriptReference/EditorGUIUtility.IconContent.html

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:235]: Well you are only resolving to one component, so shouldn't we use the singular in this case?

**Comment by Gursaanj** [comment]
> @m4rtint its the `[Resolve][SerializeField] private SomethingElse _something` line that's giving the error. The `int` field and the struct itself is set up correctly I believe

Yup exactly!! the `[Resolve]` attribute can't be put on field declared structs (i.e. `_something`) but you can put it on the individual fields declared within the struct itself (i.e. `SliderBehaviour`)

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:32]: Probably not, was going to make a button for Sibling gameObjects but not sure if the use case is used enough to justify it

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:96]: > Disable it if it's a scriptableObject?
> 
> we can leave it to iteration 2
> 
> Screen Shot 2021-12-31 at 12 57 55 PM

Hmm good point, will look into it in a couple of hours, and will let you know if we should leave it to iteration 2. Could be the same case as structs, where the fields themselves can have the attribute but the actual ScriptableObject declared field cannot

**Comment by m4rtint** [comment]
> > @m4rtint its the `[Resolve][SerializeField] private SomethingElse _something` line that's giving the error. The `int` field and the struct itself is set up correctly I believe
> 
> Yup exactly!! the `[Resolve]` attribute can't be put on field declared structs (i.e. `_something`) but you can put it on the individual fields declared within the struct itself (i.e. `SliderBehaviour`)

Ahh I see, that makes sense. 
I'm wondering instead of giving an error, u can just ignore it.

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:235]: oh in my headd, you are showing all the components in the children. but what you wrote is fine too.

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:238]: >character to flag it as a tooltip.
what the fuck, why not just build it in.

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:69]: How difficult is it to implement it like the Odin style? 
`[Resolve(IncludeInactive = false)]`

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:96]: Actually, @m4rtint  How is it supposed to work for ScriptableObjects? Since they are not Components or Gameobjects they cant just exist as a child, parent or self if that makes sense, They will always be a field of a component, instead of the actual component

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:69]: Wait im still confused, what is the difference between how it is now ,
image,

And how you want it to be?

Comment (m4rtint) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:69]: I didn't see it. Ita fine as it is then

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:96]: Nothing should happen when a [Resolve] is on the scriptableobjects

File: Assets/_BForBoss/_Utility/Scripts/PropertyAttributes/ResolveAttribute.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Scripts/PropertyAttributes/ResolveAttribute.cs:4]: Rename this to 
Perigon.Utility

**Comment by m4rtint** [comment]
Everything works great outside even after export - only issue is the name space. 
QoL for next time:
1. Would be nice if you can get this sort of error printed out for [Resolve] Struct issue, instead of printing in the log continuously 
Screen Shot 2022-01-03 at 1 30 42 AM

2. Another QoL suggestion- would be cool to enable an attribute to use `FindObjectOfType<>` 
So if you have a root manager object that needs some buttons inside another Root Canvas for example. Be able to find that. 

I'll approve when namespace is fixed.

File: Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [reply Assets/_BForBoss/_Utility/Editor/PropertyDrawers/ResolverPropertyDrawer.cs]: https://perigongames.atlassian.net/browse/BFB-153

File: Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:156]: What are these changes for?

File: Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [inline Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:168]: If I understand correctly, it finds the component within the Siblings of the parent?
A
  - B
   - - C [Resolve]
  - D
  - E
  - F

So it tries to find B, D, E F?

File: Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (m4rtint) [reply Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:156]: A
    - B
     --C [Resolve]
       --- D
          ----H
       --- E
          ----F
          ----G


So does it try to find all of D,H,E,F,G? 
Where as before it goes for only D, H?

Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:156]: Basically I realised that the way I had before did not account for children of children (which is how GetComponentsInChildren works) so it was inconsistent with the functionality when trying to resolve components instead of GameObjects. 

So in this case it would find D,H,E,F,G (like how the component one would work). This way also allows the `IncludeInactive` parameter to take into account for the gameObject search as well

File: Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs
Comment (Gursaanj) [reply Assets/_BForBoss/_Utility/Scripts/Editor/PropertyDrawers/ResolverPropertyDrawer.cs:168]: Nope no siblings only the direct trail of parents (I.e. B and A, assuming B is a child of A)

**Review by m4rtint — APPROVED** [review]

**Review by m4rtint — CHANGES_REQUESTED** [review]
The build fails currently

**Comment by unoctium1** [comment]
@Gursaanj just took a look at the branch, pretty sure the issue is that the Editor folder in Utility is missing its own asmdef file. Creating an editor only file there should fix builds

**Comment by m4rtint** [comment]
> @Gursaanj just took a look at the branch, pretty sure the issue is that the Editor folder in Utility is missing its own asmdef file. Creating an editor only file there should fix builds

I tried that, i don't think it'll work, since it still needs reference to unity.IMGUI, which i can't find the reference to.
A possible work around would be to wrap everything into 
`#if UNITY_EDITOR`

**Comment by unoctium1** [comment]
> I tried that, i don't think it'll work, since it still needs reference to unity.IMGUI, which i can't find the reference to. A possible work around would be to wrap everything into `#if UNITY_EDITOR`

I'm not sure I get what you mean, how would an edtor only asmdef be missing imgui? I tried creating an editor asmdef and didn't run into any issues, but I don't think I really understand the problem you're describing either

**Comment by m4rtint** [comment]
> > I tried that, i don't think it'll work, since it still needs reference to unity.IMGUI, which i can't find the reference to. A possible work around would be to wrap everything into `#if UNITY_EDITOR`
> 

> I'm not sure I get what you mean, how would an edtor only asmdef be missing imgui? I tried creating an editor asmdef and didn't run into any issues, but I don't think I really understand the problem you're describing either

Oh i miss understood what you meant. 
Screen Shot 2022-01-04 at 10 41 29 AM
But this is what meant right/

**Comment by unoctium1** [comment]
@m4rtint 

Yeah that's all I meant, it built when I tested locally on my end

**Review by m4rtint — APPROVED** [review]
Lgtm

**Pull request merged**
